### ISSUE TITLE
Temporal embeddings with output_metric?

### ISSUE THREAD

**Issue by MNoichl** [lmcinnes/umap#318]
TLDR: _Could the new output_metric feature be used to produce a umap-embedding with one interpretable axis?_

Thanks for the library in general, and the great maintainance!

I'm trying to improve on embeddings with a temporal axis, as discussed here:
https://github.com/lmcinnes/umap/issues/137

I was wondering if this was possible by defining an embedding space (https://umap-learn.readthedocs.io/en/0.4dev/embedding_space.html), in which one dimenson is fixed to the timestamps of the datapoints, while they can freely arrange themselves on the other dimension(s). 

Somewhat like what is happening here:
https://blockbuilder.org/MNoichl/8fa55c9b03008a1138f22be87444fe19

Is this possible, and how could one go about implementing it?

**Comment by lmcinnes** [comment]
Off the top of my head I can't think of a way to make this happen; but I've found that the ``output_metric`` options can be quite subtle, and there are often somewhat less obvious ways to make use of it to get what you want.

I'll try to give it some thought and get back to you.

**Comment by MNoichl** [comment]
Thank you for having a look! Would it help if I provided a little toy-dataset?

**Comment by lmcinnes** [comment]
Actually yes, a toy dataset (with some expected sorts of results) would be helpful. Thanks.

**Comment by MNoichl** [comment]
I hope the attached file is useful. It's citation-data of papers in philosophy and economics ('CR') and their publication-year ('PY'). I would expect a pattern of two clusters, somewhat different in structure, one for econ, and one for philosophy. The philosophy cluster can be expected to be more dense then the econ-one, as it's drawn from a smaller subfield. There will also be a tight cluster of datapoints with missing data. Basically something like this:

![image](https://user-images.githubusercontent.com/34612138/69264859-71460380-0bc8-11ea-8b7e-6b6e6d765811.png)

These should flow alongside eachother with little (although not necessarily no) interaction. And the econ cluster will probably go way longer back in time.

I hope that this is useful. If can improve it in some way, please tell me, and thank you for your time. 

The graphic was produced from the file using the following code: 
```{python}
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import umap
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import TruncatedSVD

data = pd.read_pickle('econ_phil_bib_coupling.pkl')

corps = [[str(n) for n in x] for x in data['CR']]
vec = CountVectorizer(tokenizer=lambda doc: doc, lowercase=False)
bow = vec.fit_transform(corps)
SVD = TruncatedSVD(n_components= 200, n_iter=7, random_state=42)
XSVD = SVD.fit_transform(bow)
twodmap = umap.UMAP(n_components= 2,metric="cosine",n_neighbors=20,min_dist=0.1,verbose=True).fit(XSVD)
plt.scatter(twodmap.embedding_[:,0],twodmap.embedding_[:,1], s=5, linewidth=0, c='black', alpha=0.5)
```
[econ_phil_bib_coupling.zip](https://github.com/lmcinnes/umap/files/3870600/econ_phil_bib_coupling.zip)

**Comment by lmcinnes** [comment]
So i chatted with colleagues, and we have a plan that might work. I haven't had a chance to try it. The idea is as follows:
define a custom distance with gradient function for the ``output_metric`` as 2D euclidean distance, except that the gradient for the x-coordinate is set to 0.  If you then initialize the embedding with the x-coordinates given by time, the optimization will be able to move points in the y-direction, but not the x-direction (and thus not change the time coordinate), while still measuring all the distances in the the full 2D plane.

As I said, I haven't had a chance to try it, so I have no idea how well this may or may not work, but it certainly seems like a plausible approach.

**Comment by MNoichl** [comment]
Thank you for your answer! So this what I came up with. Here is the normal umap of the dataset from above:

![Download (5)](https://user-images.githubusercontent.com/34612138/70142863-83c93e00-169a-11ea-804d-74c94b91b93b.png)

Using the following custom gradient, I got the result below (same clusters):
```{python}
@numba.njit(fastmath=True)
def adjusted_euclidean_grad(x, y):
    g = np.zeros_like(x)
    result = 0.0
    for i in range(x.shape[0]):
        result += (x[i] - (y[i])) ** 2
        g[i] = (x[i] - y[i])
    dist = np.sqrt(result) 
    gradient = g/(1e-6 + dist)
    gradient[0] = 0
    return dist, gradient

init = data.PY.values.reshape((-1,1))
init = np.append(init,np.random.rand(len(init),1), axis=1)

timed_mapper = umap.UMAP(output_metric=adjusted_euclidean_grad,
                         metric="cosine",
                         n_neighbors=25,
                         min_dist=0.01,
                         random_state=42,
                         verbose=True,
                         init=init).fit(XSVD)
```

![Download (7)](https://user-images.githubusercontent.com/34612138/70143495-11f1f400-169c-11ea-9b61-9b535c420474.png)

Is this basically what you had in mind? I think it looks already pretty good. 

One issue I ran into was, that the correct cluster seperation/consistency tends to be quite bad in larger datasets, unless I choose a very high n_neighbours (<100). But this gets computationally too expensive for large datasets. Is there another parameter that might be worth playing around with?

Thank you for your help!

**Comment by lmcinnes** [comment]
It looks like what I had in mind. In terms of performance -- the custom metrics are just slower because they have to take a different more generic code path. I don't think there are any easy ways to speed that up significantly in the short term. Sorry.

**Comment by MNoichl** [comment]
It's not that bad, I can still work well with it. 
So, thank you again for all your aid, I'll close the issue.
